### PR TITLE
#338 Create Automation for PS

### LINKED ISSUE #338: Change Requests - Automate CR Implementation for Proposed Solutions
## Description

Automatically implement changes (where possible) when a proposed solution is selected in reviewing a standard scope Change Request. This will help cut down on the amount of manual work users will need to complete when implementing a change request.

## Acceptance Criteria

When an ISSUE, DEFINITION_CHANGE, or OTHER Change Request (CR) is reviewed and consequently a proposed solution is selected as approved, do the following, depending on the whether the WBS Element attached to the CR is a:

### Project
- Add the budget impact amount from the approved proposed solution to the project's budget
- Create a change to record the increase (or decrease) in budget
### Work Package
- Add the budget impact amount from the approved proposed solution to the work package's project's budget
- Create a change to record the increase (or decrease) in budget
- Add the timeline impact amount from the proposed solution to the work package's duration
- Create a change to record the increase (or decrease) in duration

## Proposed Solution

Implement this business logic in the change request review endpoint.

## Mocks

N/A this is back-end only business logic changes.

### PR DESCRIPTION
**Pull request by Peyton-McKee**
## Changes

Automated the updating of either the project or work package for a wbs element of a change request and added changes to reflect the updates.

## Notes

Creating this PR to see if I’m on the right track
Not sure if this works at all lol

## Test Cases

- Case A
- Edge case
- ...

## Screenshots

_If you made UI changes you must post a screenshot of the whole page for each change in 1) a normal sized window and 2) the smallest possible window_

_If you did any manual testing (e.g., with Postman), put screenshots of the http request and before and after of the db_

_If none of this applies, you can delete this section_

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #338

### CODE REVIEW

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 1]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
         approved: true
       }
     });
-  }
+    const element = await prisma.WBS_Element.findUnique({
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:79]: rename to `wbs` or something

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 2]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+      where: {wbsElementId: foundCR.wbsElementId}
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:80]: you'll probably have to include objects here using `include`

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 3]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+    });
 
+
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:83]: will have to error out if (!element)

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 4]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
         approved: true
       }
     });
-  }
+    const element = await prisma.WBS_Element.findUnique({
+      where: {wbsElementId: foundCR.wbsElementId}
+    });
 
+    
+    if(!element.workPackage)
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:84]: simplify these two if's into one if

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 5]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+    {
+      if(element.project != null)
+      {
+        const updatedBudgetImpact = element.project.budgetImpact + foundPs.budgetImpact;
+        const change = {
+          changeRequestId: crId,
+          implementerId: reviewerId,
+          detail: 'Budget Impact: '.concat(element.project.budgetImpact.toString, '->', updatedBudgetImpact.toString)
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:92]: we recently merged in a helper function to generate these strings for you (buildChangeDetail)

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 6]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+        };
+        await prisma.project.update({
+          where: { projectId: element.project.projectId },
+          data: {
+            budgetImpact : updatedBudgetImpact,
+            wbsElement: {
+              update: {
+                changes: {
+                  createMany:{
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:101]: since it's only one change, doesn't need to be createmany or an array at all

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 7]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+          }
+        });
+      }
+    }
+    else
+    {
+      const updatedBudgetImpact = element.workPackage.project.budgetImpact + foundPs.budgetImpact;
+      const updatedDuration = element.workPackage.duration + foundPs.timelineImpact
+
+      const changes = [];
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:116]: can just declare it as [change1, change2]

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 8]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+        changes.push({
+          changeRequestId: crId,
+          implementerId: reviewerId,
+          detail: 'Budget Impact: '.concat(element.project.budgetImpact.toString, '->', updatedBudgetImpact.toString)
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:120]: helper function

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 9]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+        },
+        {
+          changeRequestId: crId,
+          implementerId: reviewerId,
+          detail: 'Timeline Impact: '.concat(element.project.budgetImpact.toString, '->', updatedBudgetImpact.toString)
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:125]: helper function

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 10]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+                    data: [change]
+                  }
+                }
+              }
+            }
+          }
+        });
+      }
+    }
+    else
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:111]: this might need to be an else if to ensure that the work package is not null

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 11]
```diff
@@ -76,16 +76,85 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+  }
   // update change request
   const updated = await prisma.change_Request.update({
     where: { crId },
     data: {
       reviewer: { connect: { userId: reviewerId } },
       reviewNotes,
       accepted,
-      dateReviewed: new Date()
+      dateReviewed: new Date(),
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:157]: no comma needed here

**Review by anthonybernardi — CHANGES_REQUESTED** [review]
we should write unit tests too

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 12]
```diff
@@ -76,8 +71,101 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+      return res.status(404).json({ message: `wbs element with id #${foundCR.wbsElementId} not found` });
+    }
+
+    const wp = { ...wbs.workPackage };
+
+    const proj = { ...wbs.project };
+
+    if (!wp && proj) {
+      let updatedBudgetImpact;
+      if (!proj.budget) {
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:94]: i think budget should always exist

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 13]
```diff
@@ -76,8 +71,101 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+        detail: buildChangeDetail('Budget Impact', String(proj.budget), String(updatedBudgetImpact))
+      };
+      await prisma.project.update({
+        where: { projectId: wbs.projectNumber },
+        data: {
+          budget: updatedBudgetImpact,
+          wbsElement: {
+            update: {
+              changes: {
+                createMany: {
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:111]: doesn't need to be createmany since it's just one, you can just do create

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 14]
```diff
@@ -76,8 +71,101 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+      let updatedBudgetImpact;
+      if (!proj.budget) {
+        updatedBudgetImpact = 0 + foundPs.budgetImpact;
+      } else {
+        updatedBudgetImpact = proj.budget + foundPs.budgetImpact;
+      }
+      const change = {
+        changeRequestId: crId,
+        implementerId: reviewerId,
+        detail: buildChangeDetail('Budget Impact', String(proj.budget), String(updatedBudgetImpact))
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:102]: just budget

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 15]
```diff
@@ -76,8 +71,101 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+    if (!wbs) {
+      return res.status(404).json({ message: `wbs element with id #${foundCR.wbsElementId} not found` });
+    }
+
+    const wp = { ...wbs.workPackage };
+
+    const proj = { ...wbs.project };
+
+    if (!wp && proj) {
+      let updatedBudgetImpact;
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:93]: this isn't really budget impact, if you're doing oldBudget + budgetImpact this is now the newBudget

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 16]
```diff
@@ -76,8 +71,101 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+        updatedDuration = 0 + foundPs.timelineImpact;
+      } else {
+        updatedDuration = wp.duration + foundPs.timelineImpact;
+      }
 
+      const changes = [
+        {
+          changeRequestId: crId,
+          implementerId: reviewerId,
+          detail: buildChangeDetail('Budget Impact', String(wpProj.budget), String(updatedBudgetImpact))
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:138]: just budget

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 17]
```diff
@@ -76,8 +71,101 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+        },
+        {
+          changeRequestId: crId,
+          implementerId: reviewerId,
+          detail: buildChangeDetail('Timeline Impact', String(wp.duration), String(updatedDuration))
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:143]: duration instead of timeline impact

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 18]
```diff
@@ -76,8 +71,101 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+    } else if (wp) {
+      const wpProj = await prisma.project.findUnique({
+        where: { projectId: wp.projectId }
+      });
+      if (!wpProj) {
+        return res.status(400).json({ message: 'Work package project not found' });
+      }
+      const updatedBudgetImpact = wpProj.budget + foundPs.budgetImpact;
+      let updatedDuration;
+      if (!wp.duration) {
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:128]: wp should always have duration

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 19]
```diff
@@ -76,8 +71,101 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+            update: {
+              budget: updatedBudgetImpact
+            }
+          },
+          duration: updatedDuration,
+          wbsElement: {
+            update: {
+              changes: {
+                createMany: {
+                  data: changes
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:160]: im not sure but i wonder if the project budget change should be registered on the project and not the work package

**Review by anthonybernardi — CHANGES_REQUESTED** [review]

File: src/backend/src/controllers/change-requests.controllers.ts (hunk 12 above)
Comment (Peyton-McKee) [reply src/backend/src/controllers/change-requests.controllers.ts:94]: For some reason its telling me that budget is an optional value so I just changed it to force unwrap?

File: src/backend/src/controllers/change-requests.controllers.ts (hunk 13 above)
Comment (Peyton-McKee) [reply src/backend/src/controllers/change-requests.controllers.ts:111]: It throws an error when I make it just create. Maybe because changes is an array.

File: src/backend/src/controllers/change-requests.controllers.ts (hunk 18 above)
Comment (Peyton-McKee) [reply src/backend/src/controllers/change-requests.controllers.ts:128]: This is similar to before where it says its an optional value for some reason so I changed it to a force unwrap

File: src/backend/src/controllers/change-requests.controllers.ts (hunk 19 above)
Comment (Peyton-McKee) [reply src/backend/src/controllers/change-requests.controllers.ts:160]: Do you mean instead of awaiting a work_package with a project, I should await a project with a project and do the changes there?

File: src/backend/src/controllers/change-requests.controllers.ts (hunk 12 above)
Comment (anthonybernardi) [reply src/backend/src/controllers/change-requests.controllers.ts]: hm it works on my local so not sure. and if you look at the schema budget isn't optional so idk

File: src/backend/src/controllers/change-requests.controllers.ts (hunk 13 above)
Comment (anthonybernardi) [reply src/backend/src/controllers/change-requests.controllers.ts:109]: yeah if it's just create it doesn't need to be an array, you would do like create: { change } or maybe `create: { data: { ...change}}` or something along those lines

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 20]
```diff
@@ -76,8 +71,95 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+        wbsElementId: foundCR.wbsElementId
+      }
+    });
 
+    if (!wbs) {
+      return res.status(404).json({ message: `WBS element with id #${foundCR.wbsElementId} not found` });
+    }
+    const wp = await prisma.work_Package.findUnique({
+      where: {
+        workPackageId: wbs.workPackageNumber
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:85]: workpacakge number is not an id. we have WBS numbers that are like `1.2.3` so workpackagenumber would be the 3 as part of that. the include statement (including the project and wp on the query to get the wbs) from before was much better and more efficient

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 21]
```diff
@@ -76,8 +71,95 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+      }
+    });
+    console.log('workPackage: ', wp);
+    const proj = await prisma.project.findUnique({
+      where: {
+        projectId: wbs.projectNumber
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:91]: same thing as my other comment

File: src/backend/src/controllers/change-requests.controllers.ts [hunk 22]
```diff
@@ -76,8 +71,95 @@ export const reviewChangeRequest = async (req: Request, res: Response) => {
...
+        detail: buildChangeDetail('Budget', String(proj.budget), String(newBudget))
+      };
+      await prisma.project.update({
+        where: { projectId: wbs.projectNumber },
+        data: {
+          budget: newBudget,
+          wbsElement: {
+            update: {
+              changes: {
+                createMany: {
```
Comment (anthonybernardi) [inline src/backend/src/controllers/change-requests.controllers.ts:109]: `create: change` works

**Review by anthonybernardi — CHANGES_REQUESTED** [review]
love that you wrote a bajillion tests! although they took 6 hours to finish which usually means you had some async operation that you didn't await and it kept running and running? hard to say

File: src/backend/src/controllers/change-requests.controllers.ts (hunk 19 above)
Comment (anthonybernardi) [reply src/backend/src/controllers/change-requests.controllers.ts:146]: im saying a project has a list of changes and a work package has a list of changes so which one should we put it on if we're changing the project's data but it's because of a proposed solution to the work package? i dont really know the answer tbh

**Review by anthonybernardi — CHANGES_REQUESTED** [review]
i love that you wrote so many tests but do any of them actually test your specific functionality that we're adding here?
